src/core/include/openvino/op/constant.hpp

```diff
@@ -716,6 +716,7 @@ class OPENVINO_API Constant : public Op {
     element::Type m_element_type;
     Shape m_shape{};
     std::shared_ptr<ngraph::runtime::AlignedBuffer> m_data;
+    std::shared_ptr<void> m_femgr;
```

Comment:
It is used to keep frontend manager instance is not released before ov::op::Constant which has shared memory.

Reply:
`ov::op::Constant` should be hold by `ov::Model` I don't understand why we need to duplicate frontends here

Reply:
Maybe it is duplicated, because we first resolve the segment fault for ov::op::Constant caused by dynamic frontend manager, and then found ov::Model also exit such problem, so add reference to ov::Model. I will check/verify whether can remove reference in ov::op::Constant.

Reply:
@ilya-lavrenov @ilyachur I have done some test to remove femgr reference from ov::op::constants, see test PR:https://github.com/openvinotoolkit/openvino/pull/12166
Test result shows that we still need keep this reference in ov::op::constants, some test failed, such as below 2 cases:
1) test_IENetwork.py
2) test_NGraph.py
You can reference to PR12166 CI result.

Anyhow I reproduce the same failure in local, such as test_NGraph.py

> ov2022@ov2022-Z370-AORUS-Gaming-5:~/workspace/capi/openvino/src/bindings/python/tests_compatibility$ python3 -m pytest -s test_inference_engine/test_NGraph.py -v
> ========================================================================================= test session starts ==========================================================================================
> platform linux -- Python 3.8.10, pytest-4.6.9, py-1.8.1, pluggy-0.13.0 -- /usr/bin/python3
> cachedir: .pytest_cache
> rootdir: /home/ov2022/workspace/capi/openvino/src/bindings/python
> plugins: anyio-3.5.0
> collected 6 items                                                                                                                                                                                      
> 
> test_inference_engine/test_NGraph.py::test_create_IENetwork_from_nGraph PASSED
> test_inference_engine/test_NGraph.py::test_get_IENetwork_from_nGraph PASSED
> test_inference_engine/test_NGraph.py::test_get_ops_from_IENetwork **Segmentation fault (core dumped)**

 

The failure test item is:

> def test_get_ops_from_IENetwork():
>     ie = IECore()
>     net = ie.read_network(model=test_net_xml, weights=test_net_bin)
>     func = ng.function_from_cnn(net)
>     ops = func.get_ordered_ops()
>     ops_names = [op.friendly_name for op in ops]
>     assert len(ops_names) != 0
>     assert 'data' in ops_names

It's obvious that ov::Model has been released before ops, I dump the call trace as below:

test_inference_engine/test_NGraph.py ..

> Thread 1 "python3" received signal SIGSEGV, Segmentation fault.
> 0x00007fff939b03ba in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x1266dc0) at /usr/include/c++/9/bits/shared_ptr_base.h:155
> 155                 _M_dispose();
> (gdb) bt
> #0  0x00007fff939b03ba in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x1266dc0) at /usr/include/c++/9/bits/shared_ptr_base.h:155
> #1  0x00007fff939a9b95 in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (this=0x7fffffff7908, __in_chrg=<optimized out>) at /usr/include/c++/9/bits/shared_ptr_base.h:730
> #2  0x00007fff92924f10 in std::__shared_ptr<ngraph::runtime::AlignedBuffer, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x7fffffff7900, __in_chrg=<optimized out>)
>     at /usr/include/c++/9/bits/shared_ptr_base.h:1169
> #3  0x00007fff92940612 in std::__shared_ptr<ngraph::runtime::AlignedBuffer, (__gnu_cxx::_Lock_policy)2>::operator= (this=0x1267760, __r=...) at /usr/include/c++/9/bits/shared_ptr_base.h:1265
> #4  0x00007fff929382fa in std::shared_ptr<ngraph::runtime::AlignedBuffer>::operator= (this=0x1267760, __r=...) at /usr/include/c++/9/bits/shared_ptr.h:335
> #5  0x00007fff9292f5e0 in ov::op::v0::Constant::~Constant (this=0x1267570, __in_chrg=<optimized out>) at /home/ov2022/workspace/capi/openvino/src/core/src/op/constant.cpp:243
> #6  0x00007fff923ba647 in __gnu_cxx::new_allocator<ov::op::v0::Constant>::destroy<ov::op::v0::Constant> (this=0x1267570, __p=0x1267570) at /usr/include/c++/9/ext/new_allocator.h:152
> #7  0x00007fff923b9c31 in std::allocator_traits<std::allocator<ov::op::v0::Constant> >::destroy<ov::op::v0::Constant> (__a=..., __p=0x1267570) at /usr/include/c++/9/bits/alloc_traits.h:496
> #8  0x00007fff923b916d in std::_Sp_counted_ptr_inplace<ov::op::v0::Constant, std::allocator<ov::op::v0::Constant>, (__gnu_cxx::_Lock_policy)2>::_M_dispose (this=0x1267560)
>     at /usr/include/c++/9/bits/shared_ptr_base.h:557
> #9  0x00007fff939b03c6 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x1267560) at /usr/include/c++/9/bits/shared_ptr_base.h:155
> #10 0x00007fff939a9b95 in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (this=0x7fff90685880, __in_chrg=<optimized out>) at /usr/include/c++/9/bits/shared_ptr_base.h:730
> #11 0x00007fff93a5563a in std::__shared_ptr<ov::op::v0::Constant, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x7fff90685878, __in_chrg=<optimized out>)
>     at /usr/include/c++/9/bits/shared_ptr_base.h:1169
> #12 0x00007fff93a5565a in std::shared_ptr<ov::op::v0::Constant>::~shared_ptr (this=0x7fff90685878, __in_chrg=<optimized out>) at /usr/include/c++/9/bits/shared_ptr.h:103
> #13 0x00007fff93a5ca2c in pybind11::class_<ov::op::v0::Constant, std::shared_ptr<ov::op::v0::Constant>, ov::Node>::dealloc (v_h=...)
>     at /home/ov2022/workspace/capi/openvino/src/bindings/python/thirdparty/pybind11/include/pybind11/pybind11.h:1612
> #14 0x00007fff9398f7a1 in pybind11::detail::clear_instance (self=<_pyngraph.op.Constant at remote 0x7fff90685860>)
>     at /home/ov2022/workspace/capi/openvino/src/bindings/python/thirdparty/pybind11/include/pybind11/detail/class.h:399
> #15 0x00007fff9398f8a3 in pybind11::detail::pybind11_object_dealloc (self=<_pyngraph.op.Constant at remote 0x7fff90685860>)
>     at /home/ov2022/workspace/capi/openvino/src/bindings/python/thirdparty/pybind11/include/pybind11/detail/class.h:419
> #16 0x00000000005ee3fb in _Py_Dealloc (op=<optimized out>) at ../Objects/object.c:2215
> #17 _Py_DECREF (filename=<synthetic pointer>, lineno=541, op=<optimized out>) at ../Include/object.h:478
> #18 _Py_XDECREF (op=<optimized out>) at ../Include/object.h:541
> #19 list_dealloc (op=0x7fff9060eb40) at ../Objects/listobject.c:372
> #20 0x00000000005ee8c0 in _Py_Dealloc (op=<optimized out>) at ../Objects/object.c:2215
> #21 _Py_DECREF (filename=0x88c960 "../Objects/frameobject.c", lineno=430, op=<optimized out>) at ../Include/object.h:478

Reply:
I think it is not an issue of constant op. You can have the same problem if frontend creates `ov::Tensor`. we need to fix the issue for common case, but it can require deeper refactoring of existed structures.